File: pkg/client/ts/src/orchestrator.ts
import { SuperDAWClient } from "./index.js";

export class SuperDAWOrchestrator {
    private client: SuperDAWClient;
    private daws: string[] = ["ableton", "reaper", "bitwig", "ardour", "flstudio", "logic", "cubase", "protools"];

    constructor(client: SuperDAWClient) {
        this.client = client;
    }

    async studioReset() {
        for (const daw of this.daws) {
            try {
                await this.client.transportControl(false, undefined, daw);
                await this.client.setMixer("0", 0.8, 0, daw);
            } catch (e) {
                // Ignore offline DAWs
            }
        }
    }

    async studioPlay(bpm: number = 120) {
        for (const daw of this.daws) {
            try {
                await this.client.transportControl(true, bpm, daw);
            } catch (e) { }
        }
    }

    async studioSyncTempo(bpm: number) {
        for (const daw of this.daws) {
            try {
                await this.client.transportControl(undefined, bpm, daw);
            } catch (e) { }
        }
    }
}

File: pkg/client/go/orchestrator.go
package client

import (
)

type Orchestrator struct {
	client *Client
	daws   []string
}

func NewOrchestrator(c *Client) *Orchestrator {
	return &Orchestrator{
		client: c,
		daws:   []string{"ableton", "reaper", "bitwig", "ardour", "flstudio", "logic", "cubase", "protools"},
	}
}

func (o *Orchestrator) StudioReset() {
	for _, daw := range o.daws {
		o.client.TransportControl(false, 120, daw)
		o.client.SetMixer("0", 0.8, 0, daw)
	}
}

func (o *Orchestrator) StudioPlay(bpm float64) {
	for _, daw := range o.daws {
		o.client.TransportControl(true, bpm, daw)
	}
}

func (o *Orchestrator) StudioSyncTempo(bpm float64) {
	for _, daw := range o.daws {
		o.client.TransportControl(true, bpm, daw) // Note: play=true to force sync on some daws
	}
}

func (o *Orchestrator) GetReport() map[string]string {
	report := make(map[string]string)
	for _, daw := range o.daws {
		_, err := o.client.ListClips("0", daw)
		if err == nil {
			report[daw] = "Online"
		} else {
			report[daw] = "Offline"
		}
	}
	return report
}
